### DISCUSSION TITLE
Issue using CustomStronglyTypedConnection with Package Tools

### DISCUSSION THREAD

**Discussion by amichel** [microsoft/promptflow#3629] [General]
I am trying to create a custom connection following this example:
[](https://github.com/microsoft/promptflow/blob/main/examples/tools/tool-package-quickstart/my_tool_package/tools/tool_with_custom_strong_type_connection.py)
After building and importing the tools package, I am able to create the connection from yaml that points to the tools package/module and it also works fine from a script tool, but exactly same tool method does not work when built as part of the tools package as in the above example. 
I get the following exception:

> `pf.flow.node_test failed with ResolveToolError: Tool load failed in 'Connection_Validate: (ValueTypeUnresolved) Unresolved input type 'MyCustomConnection', please check if it is supported in current version.`

I tried to work around this by calling register_connections(MyCustomConnection) inside the tool module but then I got another exception:

> pf.flow.node_test failed with ResolveToolError: Tool load failed in 'Connection_Validate': (NodeInputValidationError) Input 'connection' for node 'Connection_Validate' of type 'CustomConnection' is not supported, valid types ['MyCustomConnection'].

Any ideas welcome

**Comment by brynn-code** [comment]
@16oeahr Please help take a look at this, thanks.

**Reply by 16oeahr** [reply]
Can you share your tool package with the custom strong type connection @amichel ? I cannot repro in my local.

And please also share the details including which flow and how you run the flow.

**Reply by 16oeahr** [reply]
Please add local prompt flow package version also. If you are using the prompt flow vscode extension, please add the extension version also

**Reply by amichel** [reply]
Hi, sorry for the delay, sharing the entire repo that reproduces the problem.
[https://github.com/amichel/promptflow-tools-test](https://github.com/amichel/promptflow-tools-test)
I am using the latest stable version of the extension (v1.20.1), running everything in a devcontainer on WSL
Below is my pip list output

`Package                                  Version
---------------------------------------- -----------
aniso8601                                9.0.1
annotated-types                          0.7.0
anyio                                    4.4.0
argcomplete                              3.5.0
asttokens                                2.4.1
attrs                                    24.2.0
azure-core                               1.30.2
azure-monitor-opentelemetry-exporter     1.0.0b28
backports.tarfile                        1.2.0
blinker                                  1.8.2
certifi                                  2024.7.4
cffi                                     1.17.0
charset-normalizer                       3.3.2
click                                    8.1.7
colorama                                 0.4.6
comm                                     0.2.2
cryptography                             43.0.0
debugpy                                  1.8.3
decorator                                5.1.1
Deprecated                               1.2.14
distro                                   1.9.0
docstring_parser                         0.16
exceptiongroup                           1.2.2
executing                                2.0.1
fastapi                                  0.112.0
filelock                                 3.15.4
filetype                                 1.2.0
fixedint                                 0.1.6
Flask                                    3.0.3
Flask-Cors                               4.0.1
flask-restx                              1.3.0
gitdb                                    4.0.11
GitPython                                3.1.43
google-search-results                    2.4.1
googleapis-common-protos                 1.63.2
greenlet                                 3.0.3
h11                                      0.14.0
httpcore                                 1.0.5
httpx                                    0.27.0
idna                                     3.7
importlib_metadata                       8.0.0
importlib_resources                      6.4.0
ipykernel                                6.29.5
ipython                                  8.18.1
isodate                                  0.6.1
itsdangerous                             2.2.0
jaraco.classes                           3.4.0
jaraco.context                           5.3.0
jedi                                     0.19.1
jeepney                                  0.8.0
Jinja2                                   3.1.4
jiter                                    0.5.0
jsonschema                               4.23.0
jsonschema-specifications                2023.12.1
jupyter_client                           8.6.2
jupyter_core                             5.7.2
keyring                                  24.3.1
keyrings.alt                             5.0.1
MarkupSafe                               2.1.5
marshmallow                              3.21.3
matplotlib-inline                        0.1.7
more-itertools                           10.4.0
msrest                                   0.7.1
nest-asyncio                             1.6.0
numpy                                    2.0.1
oauthlib                                 3.2.2
openai                                   1.40.3
opentelemetry-api                        1.26.0
opentelemetry-exporter-otlp-proto-common 1.26.0
opentelemetry-exporter-otlp-proto-http   1.26.0
opentelemetry-proto                      1.26.0
opentelemetry-sdk                        1.26.0
opentelemetry-semantic-conventions       0.47b0
packaging                                24.1
pandas                                   2.2.2
parso                                    0.8.4
pexpect                                  4.9.0
pillow                                   10.4.0
pip                                      23.0.1
platformdirs                             4.2.2
prompt_toolkit                           3.0.47
promptflow                               1.14.0
promptflow-core                          1.14.0
promptflow-devkit                        1.14.0
promptflow-tools                         1.4.0
promptflow-tracing                       1.14.0
protobuf                                 4.25.4
psutil                                   5.9.8
ptyprocess                               0.7.0
pure_eval                                0.2.3
pycparser                                2.22
pydantic                                 2.8.2
pydantic_core                            2.20.1
pydash                                   7.0.7
Pygments                                 2.18.0
python-dateutil                          2.9.0.post0
python-dotenv                            1.0.1
pytz                                     2024.1
pyzmq                                    26.0.3
referencing                              0.35.1
regex                                    2024.7.24
requests                                 2.32.3
requests-oauthlib                        2.0.0
rpds-py                                  0.20.0
ruamel.yaml                              0.18.6
ruamel.yaml.clib                         0.2.8
SecretStorage                            3.3.3
setuptools                               58.1.0
six                                      1.16.0
smmap                                    5.0.1
sniffio                                  1.3.1
SQLAlchemy                               2.0.32
stack-data                               0.6.3
starlette                                0.37.2
strictyaml                               1.7.3
tabulate                                 0.9.0
test-tools                               0.0.1
tiktoken                                 0.7.0
tornado                                  6.4.1
tqdm                                     4.66.5
traitlets                                5.14.3
typing_extensions                        4.12.2
tzdata                                   2024.1
urllib3                                  2.2.2
waitress                                 2.1.2
wcwidth                                  0.2.13
Werkzeug                                 3.0.3
wheel                                    0.43.0
wrapt                                    1.16.0
zipp                                     3.19.2`

**Reply by 16oeahr** [reply]
@amichel Thanks for sharing the information. The problem lies in your package tool yaml file. The connection is not correctly written. Please pay attention to the ** part.

```yaml
test_tools.tools.test_connection.validate:
  function: validate
  inputs:
    **connection:
      custom_type:
      - TestConnection
      type:
      - CustomConnection**
  module: test_tools.tools.test_connection
  name: Test Connection Validate
  description: Validate a custom strong typed connection
  type: python
```

I am afraid current tutorial does not highlight this special config in yaml for custom strong type connection. But it provides a script in this [guide](https://microsoft.github.io/promptflow/how-to-guides/develop-a-tool/create-and-use-tool-package.html#:~:text=hello_world/yamls/hello_world_tool.yaml%3A%20Tool%20YAMLs%20defines%20the%20metadata%20of%20the%20tool.%20The%20tool%20list%20method%2C%20as%20outlined%20in%20the%20utils.py%2C%20fetches%20these%20tool%20YAMLs.) to help you auto generate the tool yaml.

**Comment by amichel** [comment]
Thanks a lot!
It solves the problem.
I actually did use the tool to generate the initial package.
I would expect the type inheritance to be resolved by reflection when
connection types are registered vs explicitly configuring custom_type/type.
